runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPipelineFuser.java

```diff
@@ -0,0 +1,222 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction.graph;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Pipeline;
+import org.apache.beam.runners.core.construction.PTransformTranslation;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PCollectionNode;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PTransformNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Fuses a {@link Pipeline} into some set of single-environment executable transforms. */
+public class GreedyPipelineFuser {
+  private static final Logger LOG = LoggerFactory.getLogger(GreedyPipelineFuser.class);
+
+  private final QueryablePipeline pipeline;
+  private final Map<CollectionConsumer, ExecutableStage> consumedCollectionsAndTransforms =
+      new HashMap<>();
+  private final Set<PTransformNode> unfusedTransforms = new HashSet<>();
+  private final Set<ExecutableStage> stages = new HashSet<>();
+
+  private GreedyPipelineFuser(Pipeline p) {
+    this.pipeline = QueryablePipeline.fromComponents(p.getComponents());
+    Collection<Set<CollectionConsumer>> rootStages = new ArrayList<>();
+    for (PTransformNode pTransformNode : pipeline.getRootTransforms()) {
+      // This will usually be a single node, either downstream of an Impulse or a Read node
+      Collection<Set<CollectionConsumer>> rootNodeStageRoots = getRootEnvTransforms(pTransformNode);
+      rootStages.addAll(rootNodeStageRoots);
+    }
+    fusePipeline(rootStages);
+  }
+
+  public static FusedPipeline fuse(Pipeline p) {
+    GreedyPipelineFuser fuser = new GreedyPipelineFuser(p);
+    return FusedPipeline.of(fuser.stages, fuser.unfusedTransforms);
+  }
+
+  private void fusePipeline(Collection<Set<CollectionConsumer>> initialConsumers) {
```

Comment:
This needs a method comment detailing what the steps of the algorithm are, and what is changing/updating in each loop and how it provides correct output.

Those comments should then appear in the body linking to the method comment.

Reply:
Done. In this context, I do not believe we meaningfully update state on each iteration - rather, we explore additional portions of the graph. I've described the parts with meaningful invariants later within the file.